### PR TITLE
gh-104549: Set __module__ on TypeAliasType

### PR DESCRIPTION
**Pull request by JelleZijlstra**
* Issue: gh-104549

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
I can't review the C code, but I much prefer this behaviour to the behaviour on `main`! Manual testing also didn't reveal any bugs.

Maybe it's also worth testing that `TypeAliasType.__module__ == 'typing'`? (I.e., accessing `__module__` on the class itself rather than instances of the class.) That's the only comment I have!

**Comment by JelleZijlstra** [comment]
@AlexWaygood unfortunately this no longer works, I get

```
test test_type_aliases failed -- Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_type_aliases.py", line 211, in test_module
    self.assertEqual(TypeAliasType.__module__, "typing")
AssertionError: <attribute '__module__' of 'typing.TypeAliasType' objects> != 'typing'
```

This seems to be an inescapable consequence of how heap types work (https://github.com/python/cpython/blob/1b5a2b085c28d230c9ef9bd9b472afc85e087ced/Objects/typeobject.c#L1045), so not sure how to get around it.

@Eclips4 good call about tests in another module, I'll add some.

As for the module being NULL, apparently that can happen in various exotic situations where there is no calling function or that function has no module. Not 100% sure how to reproduce that.

**Comment by AlexWaygood** [comment]
> This seems to be an inescapable consequence of how heap types work

That's a shame :( This inconsistency is a little unfortunate:

```pycon
>>> from typing import TypeAliasType, TypeVar
>>> TypeAliasType.__module__
<attribute '__module__' of 'typing.TypeAliasType' objects>
>>> TypeVar.__module__
'typing'
>>> type T = int | str
>>> T.__module__
'__main__'
>>> TypeVar("T").__module__
'__main__'
```

**Comment by AlexWaygood** [comment]
> That's a shame :( This inconsistency is a little unfortunate:

(But if there's no way of resolving that, then this is still definitely an improvement on the status quo :)

**Comment by JelleZijlstra** [comment]
Possible solutions:

- Make TypeAliasType into a static type. I think there is no strong reason for it to be a heap type, unlike with Generic and TypeVar; I made it a heap type for consistency with the other types in the file.
- Give it a TP_MANAGED_DICT and put the `__module__` in the dict; this is how TypeVar works, but it feels wasteful, and I don't see a need to allow setting arbitrary attributes on type aliases.

**Comment by AlexWaygood** [comment]
> * Give it a TP_MANAGED_DICT and put the `__module__` in the dict; this is how TypeVar works, but it feels wasteful, and I don't see a need to allow setting arbitrary attributes on type aliases.

Yeah, I agree that that allowing arbitrary setting of attributes seems unnecessary

**Comment by JelleZijlstra** [comment]
oops, messed up the merge

File: Modules/_typingmodule.c [hunk 1]
```diff
@@ -56,9 +57,11 @@ _typing_exec(PyObject *m)
     EXPORT_TYPE("ParamSpec", paramspec_type);
     EXPORT_TYPE("ParamSpecArgs", paramspecargs_type);
     EXPORT_TYPE("ParamSpecKwargs", paramspeckwargs_type);
-    EXPORT_TYPE("TypeAliasType", typealias_type);
     EXPORT_TYPE("Generic", generic_type);
 #undef EXPORT_TYPE
+    if (PyModule_AddObjectRef(m, "TypeAliasType", &_PyTypeAlias_Type) < 0) {
```
Comment (AlexWaygood) [inline Modules/_typingmodule.c]: Looks like there's a bunch of compiler warnings on this line

**Review by carljm — APPROVED** [review]
Assuming the refleak buildbots are happy, this LGTM.

**Pull request merged**